File: server/src/modules/tos/repository_operations/create_tos.rs
use chrono::Utc;
use sea_orm::*;
use uuid::Uuid;

use ::entity::table_of_specifications;
use crate::utils::{AppError, AppResult};

pub async fn create_tos(
    db: &DatabaseConnection,
    id: Uuid,
    class_id: Uuid,
    grading_period_number: i32,
    title: &str,
    classification_mode: &str,
    total_items: i32,
    time_unit: &str,
    easy_percentage: f64,
    medium_percentage: f64,
    hard_percentage: f64,
    remembering_percentage: f64,
    understanding_percentage: f64,
    applying_percentage: f64,
    analyzing_percentage: f64,
    evaluating_percentage: f64,
    creating_percentage: f64,
) -> AppResult<table_of_specifications::Model> {
    let now = Utc::now().naive_utc();
    let tos = table_of_specifications::ActiveModel {
        id: Set(id),
        class_id: Set(class_id),
        grading_period_number: Set(grading_period_number),
        title: Set(title.to_string()),
        classification_mode: Set(classification_mode.to_string()),
        total_items: Set(total_items),
        time_unit: Set(time_unit.to_string()),
        easy_percentage: Set(easy_percentage),
        medium_percentage: Set(medium_percentage),
        hard_percentage: Set(hard_percentage),
        remembering_percentage: Set(remembering_percentage),
        understanding_percentage: Set(understanding_percentage),
        applying_percentage: Set(applying_percentage),
        analyzing_percentage: Set(analyzing_percentage),
        evaluating_percentage: Set(evaluating_percentage),
        creating_percentage: Set(creating_percentage),
        created_at: Set(now),
        updated_at: Set(now),
        deleted_at: Set(None),
    };

    tos.insert(db)
        .await
        .map_err(|e| AppError::InternalServerError(format!("Failed to create TOS: {}", e)))
}

File: client/lib/data/repositories/tos/operations/create_tos.dart
import 'package:dartz/dartz.dart';
import 'package:likha/core/errors/failures.dart';
import 'package:likha/core/sync/mutation_result.dart';
import 'package:likha/core/sync/sync_queue.dart';
import 'package:likha/core/utils/typedef.dart';
import 'package:likha/data/datasources/local/tos/tos_local_datasource.dart';
import 'package:likha/data/models/tos/tos_model.dart';
import 'package:likha/domain/tos/entities/tos_entity.dart';
import 'package:uuid/uuid.dart';

ResultFuture<MutationResult<TableOfSpecifications>> createTos(
  TosLocalDataSource localDataSource,
  SyncQueue syncQueue, {
  required String classId,
  required Map<String, dynamic> data,
}) async {
  try {
    final tosId = const Uuid().v4();
    final queueEntryId = const Uuid().v4();
    final now = DateTime.now();

    final optimisticModel = TosModel(
      id: tosId,
      classId: classId,
      termNumber: (data['term_number'] as num?)?.toInt() ?? (data['quarter'] as num).toInt(),
      title: data['title'] as String,
      classificationMode: data['classification_mode'] as String,
      totalItems: (data['total_items'] as num).toInt(),
      timeUnit: data['time_unit'] as String? ?? 'days',
      easyPercentage: (data['easy_percentage'] as num?)?.toDouble() ?? 50.0,
      mediumPercentage: (data['medium_percentage'] as num?)?.toDouble() ?? 30.0,
      hardPercentage: (data['hard_percentage'] as num?)?.toDouble() ?? 20.0,
      createdAt: now,
      updatedAt: now,
      syncStatus: SyncStatus.pending,
    );

    final db = await localDataSource.localDatabase.database;
    await db.transaction((txn) async {
      await localDataSource.saveTos(optimisticModel, txn: txn);
      await syncQueue.enqueue(
        SyncQueueEntry(
          id: queueEntryId,
          entityType: SyncEntityType.tableOfSpecifications,
          operation: SyncOperation.create,
          payload: optimisticModel.toPayload(),
          status: SyncStatus.pending,
          retryCount: 0,
          maxRetries: 5,
          createdAt: now,
        ),
        txn: txn,
      );
    });

    return Right(MutationResult(entity: optimisticModel, status: SyncStatus.pending));
  } catch (e) {
    return Left(ServerFailure(e.toString()));
  }
}
